### PR TITLE
[HLS-3437] - helioslogrus instrumentation

### PR DESCRIPTION
**Pull request by LironKS**

### CODE REVIEW

File: proxy-libs/helioslogrus/go.mod [hunk 1]
```diff
@@ -0,0 +1,23 @@
+module github.com/helios/go-sdk/proxy-libs/helioslogrus
+
+go 1.18
+
+require (
+	github.com/sirupsen/logrus v1.9.0
+	go.opentelemetry.io/otel v1.13.0
```
Comment (nozik) [inline proxy-libs/helioslogrus/go.mod:7]: We had problems with v1.12.0, right? Don't remember where but it broke one of our integrations

File: proxy-libs/helioslogrus/helioslogrus.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	"go.opentelemetry.io/otel/trace"
+)
+
+type Hook struct {
+	levels []logrus.Level
+}
+
+var _ logrus.Hook = (*Hook)(nil)
+
+const HS_API_ENDPOINT = "https://app.gethelios.dev"
```
Comment (nozik) [inline proxy-libs/helioslogrus/helioslogrus.go:16]: make it private

File: proxy-libs/helioslogrus/helioslogrus.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	ctx := entry.Context
+	if ctx == nil {
+		return nil
+	}
+
+	span := trace.SpanFromContext(ctx)
+	if !span.IsRecording() {
+		return nil
+	}
+	entry.Data["go_to_helios"] = fmt.Sprintf("%s?actionTraceId=%s&spanId=%s", HS_API_ENDPOINT, span.SpanContext().TraceID(), span.SpanContext().SpanID())
```
Comment (nozik) [inline proxy-libs/helioslogrus/helioslogrus.go:58]: In other logs we're adding the logger type, and a timestamp (check the node and python impls)

File: proxy-libs/helioslogrus/helioslogrus_test.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+func TestLogInstrumentation(t *testing.T) {
+	sr := tracetest.NewSpanRecorder()
+	tp := sdktrace.NewTracerProvider(sdktrace.WithSpanProcessor(sr))
+	otel.SetTracerProvider(tp)
+	propagator := propagation.NewCompositeTextMapPropagator(propagation.TraceContext{}, propagation.Baggage{})
+	otel.SetTextMapPropagator(propagator)
+	tracer := tp.Tracer("TestLogrus")
+	ctx, span := tracer.Start(context.Background(), "hello-span")
+
+	l := logrus.New()
```
Comment (nozik) [inline proxy-libs/helioslogrus/helioslogrus_test.go:28]: Why aren't we patching logsrus.New?

**Review by nozik — CHANGES_REQUESTED** [review]

File: proxy-libs/helioslogrus/helioslogrus_test.go (hunk 4 above)
Comment (LironKS) [reply proxy-libs/helioslogrus/helioslogrus_test.go:28]: We can but it won't be enough cause it's possible to write log without New
We can add it  as another method to helioslogrus (and not do the whole proxy-lib thing) wdyt?

Comment (nozik) [reply proxy-libs/helioslogrus/helioslogrus_test.go:29]: True, but why not handle the case where `New` is called? The other case can be "handled" but adding docs

Comment (nozik) [reply proxy-libs/helioslogrus/helioslogrus_test.go:29]: True, but why not handle the case where New is called? The other case can be "handled" but adding docs

**Review by nozik — CHANGES_REQUESTED** [review]

File: proxy-libs/helioslogrus/helioslogrus_hook.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+	"go.opentelemetry.io/otel/trace"
+)
+
+type heliosHook struct {
+	levels []logrus.Level
+}
+
+// var _ logrus.Hook = (*Hook)(nil)
+
+const hs_api_endpoint = "https://app.gethelios.dev"
```
Comment (nozik) [inline proxy-libs/helioslogrus/helioslogrus_hook.go:17]: hsApiEndpoint

File: proxy-libs/helioslogrus/helioslogrus_hook.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+	ctx := entry.Context
+	if ctx == nil {
+		return nil
+	}
+
+	span := trace.SpanFromContext(ctx)
+	if !span.IsRecording() {
+		return nil
+	}
+	entry.Data["go_to_helios"] = fmt.Sprintf("%s?actionTraceId=%s&spanId=%s&source=logrus&timestamp=%s}", hs_api_endpoint, span.SpanContext().TraceID(), span.SpanContext().SpanID(), fmt.Sprint(time.Now().UnixNano()))
```
Comment (nozik) [inline proxy-libs/helioslogrus/helioslogrus_hook.go:59]: Should there be a `}` in the end of the formatted string?

File: proxy-libs/helioslogrus/helioslogrus_hook.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+//
+// The default is all levels between logrus.PanicLevel and logrus.WarnLevel inclusive.
+func withLevels(levels ...logrus.Level) option {
+	return func(h *heliosHook) {
+		h.levels = levels
+	}
+}
+
+// NewHook returns a logrus hook.
+func NewHook(opts ...option) *heliosHook {
```
Comment (nozik) [inline proxy-libs/helioslogrus/helioslogrus_hook.go:32]: Perhaps call it AddHeliosHook? Also, why do we need the options?

File: proxy-libs/helioslogrus/helioslogrustemp.go [hunk 8]
```diff
@@ -0,0 +1,297 @@
+package helioslogrus
+
+import (
```
Comment (nozik) [inline proxy-libs/helioslogrus/helioslogrustemp.go:3]: Why does the file name end with temp?

**Review by nozik — CHANGES_REQUESTED** [review]

File: proxy-libs/helioslogrus/helioslogrus_test.go [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+	json.Unmarshal(b.Bytes(), &data)
+
+	span.End()
+	sr.ForceFlush(ctx)
+	spans := sr.Ended()
+	assert.Equal(t, 1, len(spans))
+	value, ok := data["go_to_helios"]
+	assert.True(t, ok)
+	testedVal := fmt.Sprintf("https://app.gethelios.dev?actionTraceId=%s&spanId=%s&source=logrus", span.SpanContext().TraceID(), span.SpanContext().SpanID())
+	assert.True(t, strings.Contains(value, testedVal))
```
Comment (nozik) [inline proxy-libs/helioslogrus/helioslogrus_test.go:44]: Starts with

File: proxy-libs/helioslogrus/helioslogrus_test.go [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+	l.Out = b
+	l.WithContext(ctx).Warn("test")
+	data := map[string]string{}
+	json.Unmarshal(b.Bytes(), &data)
+
+	span.End()
+	sr.ForceFlush(ctx)
+	spans := sr.Ended()
+	assert.Equal(t, 1, len(spans))
+	value, ok := data["go_to_helios"]
```
Comment (nozik) [inline proxy-libs/helioslogrus/helioslogrus_test.go:41]: test the trace id and span id attributes as well

**Review by nozik — CHANGES_REQUESTED** [review]

**Review by nozik — APPROVED** [review]

**Pull request merged**